rs/replicated_state/src/replicated_state.rs

```diff
@@ -227,7 +227,7 @@ impl std::iter::Iterator for OutputIterator<'_> {
 pub trait PeekableOutputIterator: std::iter::Iterator<Item = RequestOrResponse> {
     /// Peeks into the iterator and returns a reference to the message that
     /// `next()` would return.
-    fn peek(&self) -> Option<&RequestOrResponse>;
+    fn peek(&self) -> Option<RequestOrResponse>;
```

Comment:
Why did we move away from a reference here?

Reply:
[Here](https://github.com/dfinity/ic/pull/1697/commits/e9f462b60e18b602ba3b2406d5caebc11f894c7c). After dozens of attempts, I finally found the right StackOverflow answer. (o:

It's actually quite nice: it uses an associated type (similar to `Iterator::Item`) so it doesn't need to leak to all users of `MessageStore` (which is still `MessageStore<T>`).